### PR TITLE
Upload image to BlobStore docker HTTP Azure function & BlobHelper

### PR DESCRIPTION
**Pull request by jananivasu**
The main checkin for the docker Azure function for uploading to Blob store and the Blob storage helper. Tests are included. Only thing remaining is the adding of AAD app secret in the key vault

### CODE REVIEW

File: UploadToBlobStoreFunction/.funcignore [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/.funcignore]: What is this file for?

File: UploadToBlobStoreFunction/.gitignore [hunk 2]
```diff
@@ -0,0 +1,42 @@
+bin
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/.gitignore]: Do we upload the gitignore file?  I can't remember

File: UploadToBlobStoreFunction/.vscode/extensions.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/.vscode/extensions.json]: I think you can delete this file.  I ran into the same thing

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+"""Blob storage operation helper"""
+import uuid
+import base64
+import io
+from applicationinsights import TelemetryClient
+
+from azure.storage.blob import BlockBlobService, PublicAccess
+
+from oauth import AutoUpdatedTokenCredential
+from settings import AZURE_STORAGE_ACCOUNT_NAME
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: I think we should standardize this, whether we're using a config file or os environment variables.  You and I are doing two different things here at the moment.  Not a priority, but just a note for future todo's

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+from settings import APPINSIGHTS_INSTRUMENTATION_KEY
+from settings import BLOB_CONTAINER_NAME
+
+
+def UploadtoBlobStore(encodeddata: bytes, filetype: str, tc: TelemetryClient) -> int:
+    """
+    Function that instantiates the telemetry client and the appropriate block blob service and calls upload helper function with that parameters
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: Long line

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+from azure.storage.blob import BlockBlobService, PublicAccess
+
+from oauth import AutoUpdatedTokenCredential
+from settings import AZURE_STORAGE_ACCOUNT_NAME
+from settings import APPINSIGHTS_INSTRUMENTATION_KEY
+from settings import BLOB_CONTAINER_NAME
+
+
+def UploadtoBlobStore(encodeddata: bytes, filetype: str, tc: TelemetryClient) -> int:
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: To my knowledge, convention for naming of python functions is lowercase with underscores... upload_to_blob_store.  We should standardize our naming conventions as a team so they're consistent across our code.

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    Function that instantiates the telemetry client and the appropriate block blob service and calls upload helper function with that parameters
+    """
+    # Create the BlockBlockService that is used to call the Blob service for the storage account
+    with AutoUpdatedTokenCredential() as token_credential:
+        block_blob_service = BlockBlobService(
+            account_name=AZURE_STORAGE_ACCOUNT_NAME, token_credential=token_credential)
+
+    container_name = BLOB_CONTAINER_NAME
+
+    retval = doActualUpload(block_blob_service,container_name,encodeddata,filetype, tc)
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: Maybe make names a little more descriptive.  Took me a second to figure out what "retval" and "tc" were.  (Unless these are standard names and I'm not aware of it).  "return_value" and "token_credential" maybe?

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    tc.flush()
+
+    return retval
+
+
+def doActualUpload(blockBlobService: BlockBlobService, containername:str, encodeddata:bytes, filetype:str, tc:TelemetryClient):
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: Is this actual upload as opposed to...test upload?

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    # Upload to blob storage
+    tc.track_trace(
+        "\nUploading to Blob storage as blob with name: " + upload_filename)
+
+     # Upload the created file, use upload_filename for the blob name
+    try:
+        blockBlobService.create_blob_from_stream(
+            containername, upload_filename, encodedImage)
+    except ValueError:
+        return 500
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: If you want, I can show you my code from CM which could help make the return HTTP responses a little more descriptive.  It helped us to include the exception in addition to the status code.

File: UploadToBlobStoreFunction/UploadToBlobStore/main.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+    tc.track_trace("Python HTTP trigger function processed a request.")
+
+    # Extract the filetype from the request
+    filetype = req.params.get('image_type')
+
+    if not filetype:
+        tc.track_trace("Could not find the file type as part of the HTTP request body")
+        tc.flush()
+        return func.HttpResponse(
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/main.py]: Ah, this is what I was referring to with the descriptive HTTP response.  *thumbs up*

File: UploadToBlobStoreFunction/UploadToBlobStore/main.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            "Could not find encoded image as part of HTTP request body",
+            status_code=400
+        )
+
+    # Upload to blob storage
+    retval = UploadtoBlobStore(encodedImage, filetype)
+
+    if retval == 200:
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/main.py]: Rename retval to response_status or status_code or something, if this is to be assigned the 200/400/500 HTTP response code.

File: UploadToBlobStoreFunction/UploadToBlobStore/main.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+    else:
+        tc.track_trace("Error occurred while uploading to blob store")
+        tc.flush()
+        return func.HttpResponse(
+            "Error occurred while uploading to blob store",
+            status_code=500
+        )
+
+def run_functriggercode(req: func.HttpRequest) -> func.HttpResponse:
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/main.py]: It looks to me like much of this code is the same as the function above...also I only see run_functriggercode used in the test file, but not in main above?

Comment (ackaufman) [reply UploadToBlobStoreFunction/UploadToBlobStore/main.py]: Beginning of a refactor?

File: UploadToBlobStoreFunction/UploadToBlobStore/settings.py [hunk 13]
```diff
@@ -0,0 +1,11 @@
+"""
```
Comment (ackaufman) [inline UploadToBlobStoreFunction/UploadToBlobStore/settings.py]: Is there anything in here that could enable someone to access any of our secrets/storage accounts if they found this file in our public repo?  Not familiar with how appinsights/AAD works.

File: UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        imageext = "jpg"
+
+        PARAMS = {'image_data': imagedata, 'image_type' : imageext}
+
+        r.params = PARAMS
+       
+        EXPECTED_BODYTEXT = "Successfully uploaded to blob store"
+        EXPECTED_STATUSCODE = 200
+    
+        resp = run_functriggercode(r)
```
Comment (seushermsft) [inline UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py]: `run_functriggercode` seems to only be used in test cases - does it have any other purpose? Ideally, your test code would run the actual code used in the HTTP trigger, but send in a mock Blob client.

File: UploadToBlobStoreFunction/UploadToBlobStore/settings.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+Configuration settings for Azure storage and AAD application ID/secret
+"""
+AZURE_STORAGE_ACCOUNT_NAME = "boulderupskillstorage"
+BLOB_CONTAINER_NAME = "staging-book-pages"
+
+APPINSIGHTS_INSTRUMENTATION_KEY = "74fcbe9f-15c5-4b15-9654-31f14ea52d18"
+
+ACTIVE_DIRECTORY_TENANT_ID = "72f988bf-86f1-41af-91ab-2d7cd011db47"
+ACTIVE_DIRECTORY_APPLICATION_ID = "a108d39e-1c4e-4a28-809f-880cbdcbb658"
+ACTIVE_DIRECTORY_APPLICATION_SECRET = "X2zKjO5aPDuZI7jSoLoL3+1BNruqXZaE7LxpSHGMqSg="
```
Comment (seushermsft) [inline UploadToBlobStoreFunction/UploadToBlobStore/settings.py]: While this is in the code, pylint will break the build and block the checkin.

File: UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+from mockito import when, mock, unstub, verify, ANY
+
+
+# To run this, run 'pytest'
+# Pytest searches for files beginning with 'test_' and calls their main function.
+# unittest.main() finds classes that inherit from 'unittest.TestCase', and run all functions whose names starts with 'test_'
+class BlobUploadTest(unittest.TestCase):
+    """Test class"""
+
+    def test_func_1(self):
```
Comment (seushermsft) [inline UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py]: Can you name this something that describes what the test does? This is the name that will show up in the build if the test breaks.

File: UploadToBlobStoreFunction/.funcignore (hunk 1 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/.funcignore]: Good question! I have no idea :) It was auto generated by the Azure core func tools I think. I can try removing and seeing if it builds fine

File: UploadToBlobStoreFunction/.gitignore (hunk 2 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/.gitignore]: Not sure either. Will check during standup tomorrow

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py (hunk 6 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: Yes, will fix this. Need to unlearn C# :)

File: UploadToBlobStoreFunction/UploadToBlobStore/settings.py (hunk 13 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/settings.py]: yes, need to move the AAD secret to key vault

File: UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py (hunk 14 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py]: hmm, will take a look. I was looking at that but then went this way :) should have persevered

File: UploadToBlobStoreFunction/.funcignore (hunk 1 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/.funcignore]: .funcignore seems to be like .gitignore but for deployment to exclude files that you do not want to be deployed. I left it as is as an empty file

File: UploadToBlobStoreFunction/.gitignore (hunk 2 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/.gitignore]: Based on searching online it looks like it is good practice to check it in so any one else who wants to work on the repository has this info too when they're working on it. So leaving it.

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py (hunk 4 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: I left the non secret ones in the config file as thats easier. moved the secrets to read from the container disk

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py (hunk 5 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: fixed

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py (hunk 6 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: fixed

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py (hunk 8 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: actual upload. but changed all the function names to be clear and consistent.

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py (hunk 7 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: done

File: UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py (hunk 9 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/BlobStorageHelper.py]: done

File: UploadToBlobStoreFunction/UploadToBlobStore/main.py (hunk 11 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/main.py]: done

File: UploadToBlobStoreFunction/UploadToBlobStore/main.py (hunk 12 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/main.py]: did a major refactor to address this

File: UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py (hunk 14 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py]: done

File: UploadToBlobStoreFunction/UploadToBlobStore/settings.py (hunk 15 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/settings.py]: moved to read from file on disk

File: UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py (hunk 16 above)
Comment (jananivasu) [reply UploadToBlobStoreFunction/UploadToBlobStore/test_UploadImage.py]: done

**Comment by jananivasu** [comment]
Closing this pull request and resubmitting a new one

**Pull request closed without merging**